File: Problems/hacker_rank/Algorithm/Arrays_and_Sorting/20150112_Insertion_Sort_Advanced_Analysis/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

long merge(int* arr, int l, int r)  {
  if ( r - l < 2 )  {
    return 0;
  }
  int i = 0;
  int m = (r + l) / 2;
  long l_cnt = merge(arr, l, m);
  long r_cnt = merge(arr, m, r);
  long cnt = l_cnt + r_cnt;
  int* tmp = (int*)malloc(sizeof(int) * (r - l));
  memset(tmp, '\0', sizeof(int) * (r - l));
  int l_idx = l, r_idx = m, t_idx = 0;
  while ( l_idx < m && r_idx < r )  {
    if ( *(arr + l_idx) <= *(arr + r_idx) ) {
      *(tmp + t_idx++) = *(arr + l_idx++);
    } else  {
      cnt += m - l_idx;
      *(tmp + t_idx++) = *(arr + r_idx++);
    }
  }
  while ( l_idx < m ) *(tmp + t_idx++) = *(arr + l_idx++);
  while ( r_idx < r ) *(tmp + t_idx++) = *(arr + r_idx++);
  //memcpy(arr + l, tmp, sizeof(int) * t_idx);
  for ( i = l; i < l + t_idx; ++i ) *(arr + i) = *(tmp + i - l);
  free(tmp);
  return cnt;
}

void merge_sort(int* arr, int n, long* ret)  {
  *ret = merge(arr, 0, n);
}

int main()  {
  int t = 0, i = 0, n = 0, j = 0;
  long ret = 0;
  scanf("%d", &t);
  for ( i = 0; i < t; ++i ) {
    scanf("%d", &n);
    int* arr = (int*) malloc(sizeof(int) * n);
    for ( j = 0; j < n; ++j ) {
      scanf("%d", arr + j);
    }
    merge_sort(arr, n, &ret);
    printf("%ld\n", ret);
    free(arr);
  }
  return 0;
}

File: Problems/hacker_rank/Algorithm/Arrays_and_Sorting/20150112_Insertion_Sort_Advanced_Analysis/Solution.scala
object Solution {
  def mergeSort(arr: Array[Int]): (Array[Int], Int) = {
    val l = arr.length
    if (l < 2)
      return (arr, 0)
    val m = l / 2
    var (left, lCnt) = mergeSort(arr.slice(0, m))
    var (right, rCnt) = mergeSort(arr.slice(m, l))
    var cnt = lCnt + rCnt
    var res: Array[Int] = Array()
    println("left\t" + left.mkString(" ") + "\tright\t" + right.mkString(" "))
    var (idx, lIdx, rIdx) = (0, 0, 0)
    while ( 0 < left.length && 0 < right.length ) {
      if ( left(0) <= right(0) )  {
        res(idx) = left(0)
        //res = res :+ left(0)
        left = left.drop(1)
      } else  {
        cnt += left.length
        res(idx) = right(0)
        //res = res :+ right(0)
        right = right.drop(1)
      }
      idx += 1
    }
    res = res.dropRight(res.length - idx)
    println(res.mkString(" ") + "\t[" + cnt + "]")
    (res ++ left ++ right, cnt)
  }
  def mergeSort2(arr: Array[Int]): (Array[Int], BigInt) = {
    val l = arr.length
    if (l < 2)
      return (arr, 0)
    val m = l / 2
    var (left, lCnt) = mergeSort2(arr.slice(0, m))
    var (right, rCnt) = mergeSort2(arr.slice(m, l))
    var cnt: BigInt = lCnt + rCnt
    var res: Array[Int] = Array.fill[Int](left.length + right.length)(0)
    var (idx, lIdx, rIdx) = (0, 0, 0)
    while ( lIdx < left.length && rIdx < right.length ) {
      if ( left(lIdx) <= right(rIdx) ) {
        res(idx) = left(lIdx)
        lIdx += 1
      } else  {
        cnt += left.length - lIdx
        res(idx) = right(rIdx)
        rIdx += 1
      }
      idx += 1
    }
    while ( lIdx < left.length )  {
      res(idx) = left(lIdx)
      idx += 1
      lIdx += 1
    }
    while ( rIdx < right.length )  {
      res(idx) = right(rIdx)
      idx += 1
      rIdx += 1
    }
    (res, cnt)
  }

  def main(args: Array[String]) {
    val inps = io.Source.stdin.getLines.toArray
    for ( i <- Range(1, inps(0).toInt * 2, 2) )  {
      val n = inps(i).toInt
      val arr = inps(i + 1).split(" ").map((s: String) => s.toInt).toArray
      println(mergeSort2(arr)._2)
    }
  }
}
